src/StringTools/InternableString.cs

```diff
@@ -265,6 +265,14 @@ public unsafe string ExpensiveConvertToString()
                         }
                     }
                 }
+
+                // The invariant that Length is the sum of span lengths is critical in this unsafe method.
+                // Violating it may lead to memory corruption and, since this code tends to run under a lock,
+                // to hangs caused by the lock getting orphaned.
+                if (destPtr != resultPtr + Length)
+                {
+                    throw new InvalidOperationException("Length property does not match sum of span lengths");
+                }
```

Comment:
The comment makes sense to me but I'm not sure I understand how the new exception would help the situation. Can you expand on that some?

Reply:
I am assuming that a length mismatch does not always mean memory corruption and a memory corruption does not always mean immediate crash. So the new code will execute and we'll get dumps (or other artifacts, possibly including a repro) at the stack where the problem happens. As opposed to getting hangs on another thread later, which is the current unfortunate situation.

Reply:
```suggestion
                // The invariant that Length is the sum of span lengths is critical in this unsafe method.
                // Violating it may lead to memory corruption and, since this code tends to run under a lock,
                // to hangs caused by the lock getting orphaned. Attempt to detect that and throw now, 
                // before the corruption causes further problems.
                if (destPtr != resultPtr + Length)
                {
                    throw new InvalidOperationException("Length property does not match sum of span lengths");
                }
```

?

Reply:
Applied, thank you!